libraries/botframework-streaming/src/webSocket/nodeWebSocket.ts

```diff
@@ -1,119 +1,138 @@
-        req.on('upgrade', function(res, socket, head): void {
-            this.watershedShed.connect(res, socket, head, wskey);
-        });
-
-        this.connected = true;
-
-        return new Promise<void>((resolve, reject): void => {
-            req.on('close', resolve);
-            req.on('error', reject);
-        });
-    }
-
-    /**
-     * Set the handler for text and binary messages received on the socket.
-     */
-    public setOnMessageHandler(handler: (x: any) => void): void {
-        this.waterShedSocket.on('text', handler);
-        this.waterShedSocket.on('binary', handler);
-    }
-
-    /**
-     * Close the socket.
-     */
-    public close(): any {
-        this.connected = false;
-
-        return this.waterShedSocket.end();
-    }
-
-    /**
-     * Set the callback to call when encountering socket closures.
-     */
-    public setOnCloseHandler(handler: (x: any) => void): void {
-        this.waterShedSocket.on('end', handler);
-    }
-
-    /**
-     * Set the callback to call when encountering errors.
-     */
-    public setOnErrorHandler(handler: (x: any) => void): void {
-        this.waterShedSocket.on('error', (error): void => { if (error) { handler(error); } });
-    }
-}
+/**
+ * @module botframework-streaming
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import * as crypto from 'crypto';
+import { IncomingMessage, request } from 'http';
+import { Socket } from 'net';
+import * as WebSocket from 'ws';
+
+import { ISocket } from '../interfaces';
+const NONCE_LENGTH = 16;
+
+export class NodeWebSocket implements ISocket {
+    private wsSocket: WebSocket;
+    private connected: boolean;
+    protected wsServer: WebSocket.Server;
+
+    /**
+     * Creates a new instance of the [NodeWebSocket](xref:botframework-streaming.NodeWebSocket) class.
+     *
+     * @param socket The `ws` WebSocket instance to build this connection on.
+     */
+    public constructor(wsSocket?: WebSocket) {
+        this.wsSocket = wsSocket;
+        this.connected = !!wsSocket;
+        this.wsServer = new WebSocket.Server({ noServer: true });
+    }
+
+    /**
+     * Create and set a `ws` WebSocket with an HTTP Request, Socket and Buffer.
+     * @param req IncomingMessage
+     * @param socket Socket
+     * @param head Buffer
+     */
+    public async create(req: IncomingMessage, socket: Socket, head: Buffer): Promise<void> {
+        return new Promise<void>((resolve, reject) => {
+            try {
+                this.wsServer.handleUpgrade(req, socket, head, (websocket) => {
+                    this.wsSocket = websocket;
+                    this.connected = true;
+                    resolve();
+                });
+            } catch (err) {
+                reject(err);
+            }
+        });
+    }
+
+    /**
+     * True if the socket is currently connected.
+     */
+    public get isConnected(): boolean {
+        return this.connected;
```

Comment:
We update the connected member only when we close. If there is an externally caused disconnection, are we setting this somehow? Could the connected flag be computed from the current state socket to reflect more precisely whether the socket is actually connected?

Reply:
This ties in well to the conversation [here](https://github.com/microsoft/botbuilder-js/pull/1425#discussion_r347525299).

I've addressed this in commit [`d3bbf1f`](https://github.com/microsoft/botbuilder-js/pull/1425/commits/d3bbf1ff49237e9548c57ec8f4d8fd888f92e58c) where it instead checks [`WebSocket.readyState`](https://github.com/websockets/ws/blob/master/lib/websocket.js#L46) of the `ws` WebSocket.